plan/eliminate_projection.go

```diff
@@ -14,50 +14,245 @@
 	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
 )
 
-// EliminateProjection eliminates projection operator to avoid the cost of memory copy in the iterator of projection.
-func EliminateProjection(p PhysicalPlan) PhysicalPlan {
-	switch plan := p.(type) {
-	case *Projection:
-		if !projectionCanBeEliminated(plan) {
-			break
-		}
-		child := plan.children[0].(PhysicalPlan)
-		child.SetSchema(plan.Schema())
-		RemovePlan(p)
-		p = EliminateProjection(child)
+// canProjectionBeEliminatedLoose checks whether a projection can be eliminated, returns true if:
+// 1. the projection changes the order of its child's output columns, or:
+// 2. the projection just copy its child's output but set a different name for every column.
+func canProjectionBeEliminatedLoose(p *Projection) bool {
+	child := p.Children()[0]
+	if p.Schema().Len() != child.Schema().Len() {
+		return false
 	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
+	for _, expr := range p.Exprs {
+		_, ok := expr.(*expression.Column)
+		if !ok {
+			return false
+		}
 	}
-	p.SetChildren(children...)
-	return p
+	return true
 }
 
-// projectionCanBeEliminated checks if a PROJECTION operator can be eliminated.
-// PROJECTION operator can be eliminated when meet the following conditions at the same time:
-// 1. fields of PROJECTION are all columns
-// 2. fields of PROJECTION are just the same as the schema of the child operator (including order, amount, etc.).
-// expressions like following cases can not be eliminated:
-// "SELECT b, a FROM t",
-// or "SELECT c AS a, c AS b FROM t WHERE d = 1",
-// or "SELECT t1.a, t2.b, t1.b, t2.a FROM t1, t2 WHERE t1.a < 0 AND t2.b > 0".
-func projectionCanBeEliminated(p *Projection) bool {
-	child := p.children[0].(PhysicalPlan)
+// canProjectionBeEliminatedStrict checks whether a projection can be eliminated, returns true if:
+// 1. the projection just copy its child's output but set a different name for every column.
+// Unlike canProjectionBeEliminatedLoose, it will return false when the projection changes the order
+// of its child's output columns.
+func canProjectionBeEliminatedStrict(p *Projection) bool {
+	child := p.Children()[0]
 	if p.Schema().Len() != child.Schema().Len() {
 		return false
 	}
 	for i, expr := range p.Exprs {
 		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		if !ok || !col.Equal(child.Schema().Columns[i], nil) {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
```

Comment:
As previous discussion, function `expression.ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression)` is too heavy and inefficient: 
 1. for every `ScalarFunction` node in the expression tree, this function will do a copy for it.
 2. uses a loop to find whether a column exists in `schema *Schema`

I prefer not to use it